src/box/lua/net_replicaset.lua

```diff
@@ -0,0 +1,268 @@
+local fiber = require('fiber')
+local net_box = require('net.box')
+
+-- The constant is copy-paste from tarantool/src/lua/socket.lua.
+local TIMEOUT_INFINITY = 500 * 365 * 86400
+
+-- Raise a net.replicaset specific error.
+local function rs_error(err_code, replicaset_name, payload)
+    local args = payload and table.copy(payload) or {}
+    args.code = err_code
+    args.replicaset = replicaset_name
+    box.error(args, 2)
+end
+
+local replicaset_methods = {}
+
+local replicaset_mt = {
+    __index = replicaset_methods,
+}
+
+-- Check that the arguments table comply table template.
+-- Template key-value pairs show possible options and their required types -
+-- a string concatenated from several types with reasonable delimiter.
+local function check_options(tbl, tmpl, name)
+    if type(tbl) ~= 'table' then
+        error(name .. ' expected to be a map')
+    end
+    for k, v in pairs(tmpl) do
+        if not string.find(v, type(tbl[k])) then
+            local fmt = 'Option \'%s\' of %s is \'%s\' while must be \'%s\''
+            error(string.format(fmt, k, name, type(tbl[k]), v))
+        end
+    end
+    for k in pairs(tbl) do
+        if tmpl[k] == nil then
+            error('Unexpected option ' .. k .. ' in ' .. name, 2)
+        end
+    end
+end
+
+-- Check that the first argument is a valid replicaset.
+local function replicaset_check(replicaset, method)
+    if getmetatable(replicaset) ~= replicaset_mt then
+        local fmt = 'Use replicaset:%s(...) instead of replicaset.%s(...)'
+        error(string.format(fmt, method, method))
+    end
+end
+
+-- Wait for the leader of replicaset to be known, limited with given timeout.
+-- The leader is the only known RW replica.
+-- Note1. This method does not wait for all peers to connect and send status.
+-- For example it will successfully return if one replica is known as RW while
+-- two others have not send status event. This is done for availability of
+-- replicaset connection: if one or more replica instance are down it should
+-- be possible to proceed requests with replicaset remains.
+-- Note2. Due to asynchronous nature of replicaset connection it is possible
+-- that the known leader is not actually a leader or perhaps not even RW.
+-- Returns remaining timeout.
+function replicaset_methods:_wait_leader(timeout)
```

Comment:
Btw, if a method is fully internal (never even used outside of this Lua file), then you can just declare it as a function and it doesn't have to be in the methods table. Then you would have to pass the `self` argument explicitly, but the function wouldn't be visible outside at all.

Up to you. Just a note.

Reply:
Oh, there were too many battles about how object's methods should be implemented.. I would prefer not to change it again.